src/lib/README.md

```diff
@@ -0,0 +1,105 @@
+Impress.js Libraries
+====================
+
+The `src/lib/*.js` files contain library functions. The main difference to plugins is that:
+
+1. Libraries are closer to the impress.js core than plugins (arguably a subjective metric)
+2. Libraries are common utility functions used by many plugins
```

Comment:
Is a library API really necessary to be provided by impress.js? It looks like this is moving away of the purpose of impress.js which is serving as an engine for presentations and being more of a container for utilities.

What's exactly the use case?

Reply:
As far as I understand this API is part of impress.js (as a whole).
These are common utility functions used both by impress.js core (some of them are directly extracted from impress.js code) and others are shared by many plugins.

I believe it's worth keeping them as part of impress.js to make plugin development easier - not to expect plugins will re-implement same things or rely on other libraries to provide this.

Reply:
The code you commented is related to the README implementing a way for users to create snippets inside impress.js namespace through the `window.impress.addLibraryFactory` function. That's not part of impress.js and is being added here.

If you're referring to the libraries that are already internal to impress.js such as `arrayify`, `byId`, `$$` and `$`: what's the point on exposing them for consumers of impress.js? Is it just to make them not to type `[...arguments]`, `document.getElementById`, `document.querySelectorAll` and `document.querySelector`?

We're just obfuscating code.

Reply:
Leaving the point about usefulness of util functions like `$`, `$$`, `byId`, etc aside.

I think we can agree that if there there is a need to share some functionality and utils between plugins. At least if it is the 'garbage collector' that allows plugins to be correctly destroyed and clean after themselves.

To make it possible for such libraries to register themselves in impress.js some new method is needed to add them. Otherwise impress.js itself will need to be aware what libs to add, which is obviously not really scalable.

I understand the need for `addLibraryFactory`. If we worry about this being exposed as public API we may 'hide' it and keep private until we think it can be used publicly.
To make impress.js more extensible it needs means (API functions) that add new functionality to it. In such regard `addLibraryFactory` is a part of impress.js core and valid addition to the API.

Reply:
> If you're referring to the libraries that are already internal to impress.js such as `arrayify`, `byId`, `$$` and `$`: what's the point on exposing them for consumers of impress.js? Is it just to make them not to type `[...arguments]`, `document.getElementById`, `document.querySelectorAll` and `document.querySelector`?

Yes, in today browsers we probably don't need utils that and we probably don't need to expose them as public library utils. I understand that they are simply utility functions of current impress.js core moved to the library.

I believe removing them now may affect rest of the plugins @henrikingo has ready to be merged, so again - I'm fine with merging it as it is and reviewing the bigger picture later to see what libs or utils can be removed in favour of native functions.

Reply:
> I believe removing them now may affect rest of the plugins @henrikingo has ready to be merged, so again - I'm fine with merging it as it is and reviewing the bigger picture later to see what libs or utils can be removed in favour of native functions.

That's cool, it's just that if we leave it exposed we might need to support it as more people start using it.

> If we worry about this being exposed as public API we may 'hide' it and keep private until we think it can be used publicly.

That looks like a good approach.